src/runtime_src/core/edge/user/device_linux.cpp

```diff
@@ -456,6 +456,82 @@ struct aie_reg_read
   }
 };
 
+struct aie_get_freq
+{
+  using result_type = query::aie_get_freq::result_type;
+
+  static result_type
+  get(const xrt_core::device* device, key_type key, const boost::any& partition_id)
+  {
+    auto dev = get_edgedev(device);
+    result_type freq = 0;
+#ifdef XRT_ENABLE_AIE
+#ifndef __AIESIM__
+    const static std::string AIE_TAG = "aie_metadata";
+    const static std::string ZOCL_DEVICE = "/dev/dri/renderD128";
+    std::string err;
+    std::string value;
+
+    // Reading the aie_metadata sysfs.
+    dev->sysfs_get(AIE_TAG, err, value);
+    if (!err.empty())
+      throw xrt_core::query::sysfs_error
+      (err + ", The loading xclbin acceleration image doesn't use the Artificial "
+       + "Intelligent Engines (AIE). No action will be performed.");
+
+    int mKernelFD = open(ZOCL_DEVICE.c_str(), O_RDWR);
+    if (!mKernelFD)
+      throw xrt_core::error(-EINVAL, boost::str(boost::format("Cannot open %s") % ZOCL_DEVICE));
+
+    uint32_t part_id = boost::any_cast<uint32_t>(partition_id);
+    struct drm_zocl_aie_freq_scale aie_arg = { part_id, freq, 0 };
+    if (ioctl(mKernelFD, DRM_IOCTL_ZOCL_AIE_FREQSCALE, &aie_arg)) {
+      throw xrt_core::error(-errno, boost::str(boost::format("Reading frequency from AIE partition %d failed") % part_id));
+    }
+#endif
+#endif
+    return freq;
+  }
+};
+
+struct aie_set_freq_req
+{
+  using result_type = query::aie_set_freq_req::result_type;
+
+  static result_type
+  get(const xrt_core::device* device, key_type key, const boost::any& partition_id, const boost::any& freq)
+  {
+    auto dev = get_edgedev(device);
+#ifdef XRT_ENABLE_AIE
+#ifndef __AIESIM__
+    const static std::string AIE_TAG = "aie_metadata";
+    const static std::string ZOCL_DEVICE = "/dev/dri/renderD128";
+    std::string err;
+    std::string value;
+
+    // Reading the aie_metadata sysfs.
+    dev->sysfs_get(AIE_TAG, err, value);
+    if (!err.empty())
+      throw xrt_core::query::sysfs_error
+      (err + ", The loading xclbin acceleration image doesn't use the Artificial "
+       + "Intelligent Engines (AIE). No action will be performed.");
+
+    int mKernelFD = open(ZOCL_DEVICE.c_str(), O_RDWR);
+    if (!mKernelFD)
+      throw xrt_core::error(-EINVAL, boost::str(boost::format("Cannot open %s") % ZOCL_DEVICE));
```

Comment:
This looks like code duplication, please find a way to share in separate function.